Add Helm schema

ubergesundheit opened the pull request:
This PR adds a helm value schema and the means to generate it from the default values.

The changes are mostly in the `values.yaml.tmpl`. Type annotations were added to make sure the `values.schema.json` file does not contain any `"type": "null"` fields.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.


<!-- Description of change -->

Fixes: https://github.com/cilium/cilium/issues/22831

```release-note
Add helm values.schema.json file for validating supplied values for correct type.
```

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -401,30 +401,30 @@ bpf:
   preallocateMaps: false
   # -- (int) Configure the maximum number of entries in auth map.
   # @default -- `524288`
```
Did you investigate why previously those were set to `nil` with the default value manually documented?

Since it would have been way simpler to just set it to a default value, I do wonder why we previously opted to set it to nil.

ubergesundheit replied:
I guess they we're left at nil because the defaults are being enforced by cilium itself so my guess is that devs wanted to show that is possible to configure it. Another way would've been to put it in the values.yaml and to comment it out.

gandro replied:
I checked git blame and most were indeed uncommitted to ensure they get picked up by Helm Docs: https://github.com/cilium/cilium/pull/21650

What remains to be discussed is there is any observable change due the value now being present in the configmap (e.g. with regards to downgrades). Let's see what the upgrade/downgrade test in CI says.

chancez commented on Makefile.defs:
```diff
@@ -200,6 +200,10 @@ YQ_VERSION ?= "4.40.5"
 
+HELM_SCHEMA_GEN_VERSION ?= "3.14.0"
+HELM_SCHEMA_GEN_SHA ?= "e3d2a30717adcaf94dc68ae8f2940db875d235fc7e59f62091c2a46ec1adc3d1"
+HELM_SCHEMA_GEN_IMAGE ?= "quay.io/geraldpape/docker-helm:$(HELM_SCHEMA_GEN_VERSION)@sha256:$(HELM_SCHEMA_GEN_SHA)"
```
https://github.com/karuppiah7890/helm-schema-gen/blob/main/cmd/root.go#L38-L41
The logic for this is 3 lines of actual code. If we do want this, IMO we should just hand vendor that into `tools/` and `go run` it instead.

joestringer replied:
I agree with Chance, it does not make sense to rely on a third-party docker image built from an unpopular third-party github repository like this.

ubergesundheit replied:
Updated to quay.io/cilium/helm-toolbox:v1.1.0 now

chancez reviewed (commented):
This looks like it could be promising, but I have a few concerns:

- This could easily become an accidental breaking change if we're not careful. Is there a way to bypass Helm schema validation? I'm not seeing a flag for it. (`--disable-openapi-validation` is for k8s validation, not values.yaml validation).
- I believe the chart has a number of configuration options that do not have any defaults defined in `values.yaml`. `git grep hasKey -- install/kubernetes` for some examples. If they're not defined in the schema, would these options be disallowed? If so: all of these would need to be added to the `values.yaml.tmpl`/`values.yaml` in order to be correctly added to the schema.
- What if there are cases where auto-generation doesn't work for some case, or what should we do if we want to add extra validation rules? I like the idea of auto-generating the schema, but I think it would be useful to plan a bit ahead for how we might maintain this longer term.
- I made this as a line comment, but the actual logic to generate the schema seems simple enough that we don't need to add a new tools image, but could probably write a simple implementation in `tools/`.

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -401,30 +401,30 @@ bpf:
   # table.
   # @default -- `524288`
-  ctTcpMax: ~
+  ctTcpMax: 524288
```
These map size changes are incorrect. By default, Cilium picks map sizes based on available memory. We deliberately omit the configuration because there's no way to specify one configuration that matches all deployments. Please remove the map reconfiguration from this PR.

(Consequences for this specific change could be that it causes significant packet loss during upgrade)

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -45,11 +45,11 @@ k8sServicePort: ""
   # The rate limiter will allow short bursts with a higher rate.
   # @default -- 10 for k8s up to 1.26. 20 for k8s version 1.27+
-  burst:
+  burst: 20
```
I think this is being dynamically selected by the cilium-agent, so hardcoding a default in the Helm chart is working against the desired configuration (based on the local environment).

joestringer reviewed (changes requested):
Thanks for working on this. I think the idea here is interesting, but we need to be super careful if we are changing defaults. I highlighted at least one case below where this could cause impact to production systems. More broadly for the other settings which are undefined by default (where the setting is left up to the cilium-agent configuration), I'm not sure whether we are ready to specify defaults inside Helm charts for every one of these, or how those settings may themselves impact upgrade. Another aspect to it is that the cilium-agent can more easily adapt to local environment requirements and pick good defaults, so by hardcoding actual values into the Helm charts, this can make Cilium less adaptable.

ubergesundheit committed:
helm: Introduce values.schema.json and tooling

Auto-generate a values.schema.json file from type annotations in
values.yaml.tmpl.

Update the helm-toolbox container image to add additional tooling.

Fixes cilium#22831

Signed-off-by: Gerald Pape <gerald@giantswarm.io>

joestringer commented on Makefile.defs:
```diff
@@ -192,9 +192,9 @@ GO_CLEAN = $(GO) clean $(GO_CLEAN_FLAGS)
-HELM_TOOLBOX_IMAGE ?= "quay.io/cilium/helm-toolbox:$(HELM_TOOLBOX_VERSION)@sha256:$(HELM_TOOLBOX_SHA)"
+HELM_TOOLBOX_VERSION ?= "v1.1.0"
+HELM_TOOLBOX_SHA ?= "93ec4d691c634b10a89f1db5d88faaae34318465e81c3bff576256db4aa264c2"
+HELM_TOOLBOX_IMAGE ?= "quay.io/geraldpape/helm-toolbox:$(HELM_TOOLBOX_VERSION)@sha256:$(HELM_TOOLBOX_SHA)"
```
We can coordinate this with https://github.com/cilium/helm-toolbox/pull/3. Once that PR is merged, we can tag a new cilium/helm-toolbox and update these references.

ubergesundheit replied:
Yes, that would be awesome

joestringer replied:
I'm trying this out locally via the PR, and a bunch of the flags like `--omit-additional-properties --omit-default --omit-description --omit-title --omit-required-properties` are not available. Are those net-new features?

This is `--help` with version 0.9.0:

```
helm-schema automatically generates a jsonschema file for helm charts from values files

Usage:
  helm-schema [flags]

Flags:
  -c, --chart-search-root string       directory to search recursively within for charts (default ".")
  -x, --dont-strip-helm-docs-prefix    Disable the removal of the helm-docs prefix (--)
  -d, --dry-run                        don't actually create files just print to stdout passed
  -h, --help                           help for helm-schema
  -s, --keep-full-comment              Keep the whole leading comment (default: cut at empty line)
  -l, --log-level string               Level of logs that should printed, one of (panic, fatal, error, warning, info, debug, trace) (default "info")
  -n, --no-dependencies                don't analyze dependencies
  -o, --output-file string             jsonschema file path relative to each chart directory to which jsonschema will be written (default "values.schema.json")
  -k, --skip-auto-generation strings   comma separated list of fields to skip from being created by default (possible: title, description, required, default, additionalProperties)
  -f, --value-files strings            filenames to check for chart values (default [values.yaml])
  -v, --version                        version for helm-schema
```

ubergesundheit replied:
Yeah this changed. Try with `--skip-auto-generation title,description,required,default,additionalProperties`

ubergesundheit replied:
Pushed an update to the Makefile now

joestringer replied:
I just tagged `quay.io/cilium/helm-toolbox:v1.1.0` at sha `961693f182b9b456ed90e5274ac5df81e4af4343104e252666959cdf9570ce9e`.

joestringer commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2653,10 +2900,16 @@ clustermesh:
       # For GKE LoadBalancer, use annotation cloud.google.com/load-balancer-type: "Internal"
       # For EKS LoadBalancer, use annotation service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
       annotations: {}
+      # @schema
+      # enum: [Local, Cluster]
+      # @schema
       # -- The externalTrafficPolicy of service used for apiserver access.
-      externalTrafficPolicy:
+      externalTrafficPolicy: Cluster
+      # @schema
+      # enum: [Local, Cluster]
+      # @schema
       # -- The internalTrafficPolicy of service used for apiserver access.
-      internalTrafficPolicy:
+      internalTrafficPolicy: Cluster
```
There's a few changes here still that seem to be modifying the defaults.. though I don't know whether these should have the value `~` to make them null, or adopt these new specific defaults.

joestringer replied:
^ Probably @cilium/sig-k8s would have a better understanding of this.

ubergesundheit replied:
I wanted to avoid another case of `type: [null, string]` here.

`externalTrafficPolicy` and `internalTrafficPolicy` are being defaulted to `Cluster` if not specified. ([See k8s API spec page](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#servicespec-v1-core))

> InternalTrafficPolicy describes how nodes distribute service traffic they receive on the ClusterIP. If set to "Local", the proxy will assume that pods only want to talk to endpoints of the service on the same node as the pod, dropping the traffic if there are no local endpoints. **The default value, "Cluster"**, uses the standard behavior of routing to all endpoints evenly (possibly modified by topology and other features).

> externalTrafficPolicy describes how nodes distribute service traffic they receive on one of the Service's "externally-facing" addresses (NodePorts, ExternalIPs, and LoadBalancer IPs). If set to "Local", the proxy will configure the service in a way that assumes that external load balancers will take care of balancing the service traffic between nodes, and so each node will deliver traffic only to the node-local endpoints of the service, without masquerading the client source IP. (Traffic mistakenly sent to a node with no endpoints will be dropped.) **The default value, "Cluster"**, uses the standard behavior of routing to all endpoints evenly (possibly modified by topology and other features). Note that traffic sent to an External IP or LoadBalancer IP from within the cluster will always get "Cluster" semantics, but clients sending to a NodePort from within the cluster may need to take traffic policy into account when picking a node.

joestringer commented on Documentation/helm-values.rst:
```diff
@@ -2050,8 +2050,8 @@
      - Configure image pull secrets for pulling container images
-     - string
-     - ``nil``
+     - list
+     - ``[]``
```
Is this an incompatible change? Users may have configured this setting already with a direct string, so changing the type like this could break their existing configuration..?

joestringer replied:
Maybe if we think this is the right type for this variable moving forward, we can just document this in the upgrade notes `Documentation/operations/upgrade.rst` that the setting is changing the value type. (Same for any others that we consciously change)

ubergesundheit replied:
Not really sure if this would work at all. I suspect helm erroring if people actually specify a string here.

joestringer replied:
Huh, I see what you mean, the API reference specifies this an an array. I'm confused how we ended up with this broken option in the Helm charts in the first place. Well, thanks for the fix :)

ubergesundheit replied:
> I suspect helm erroring if people actually specify a string here.

Even before this PR

joestringer reviewed (commented):
I'll defer to others like @gandro on the general feedback here but I'll just point out a couple of the practical points below that may need a closer look. This newer proposal seems less invasive than the earlier one, so less likely to have an impact on existing users.

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -48,7 +49,7 @@ update-chart: $(ROOT_DIR)/VERSION # Update chart versions to point to the curren
 	$(QUIET)$(SED) -i 's;icon:.*;icon: $(LOGO_BASE_URL)/cilium@$(CILIUM_BRANCH)/$(LOGO_PATH);' $(CHART_FILE)
 
-check-values-yaml:
+check-values-yaml: cilium/values.schema.json
```
This target creates a dependency on generating the schema, but where is the schema used to validate the values?

ubergesundheit replied:
The schema is used by helm when installing

joestringer replied:
:+1: looks like it's supposed to be picked up by `helm lint` too, which we already run as part of the CI.

joestringer reviewed (approved):
My feedback was addressed. Thanks for the contribution!

joestringer commented on Makefile.defs:
```diff
@@ -192,8 +192,8 @@ GO_CLEAN = $(GO) clean $(GO_CLEAN_FLAGS)
 
-HELM_TOOLBOX_VERSION ?= "v1.0.2"
-HELM_TOOLBOX_SHA ?= "c5a728e61e1f04580c1a0074f9bf3ab96f7beca687385878b403bb0e52163d34"
+HELM_TOOLBOX_VERSION ?= "v1.1.0"
+HELM_TOOLBOX_SHA ?= "961693f182b9b456ed90e5274ac5df81e4af4343104e252666959cdf9570ce9e"
```
Heads up I just pushed `v1.1.1` to fix the CVEs in the images. Feel free to either bump it in this PR or I'll bump this later in another PR. https://quay.io/repository/cilium/helm-toolbox/manifest/sha256:722901c26a9edb75a514927f9f759302a0dc03edacadc835a76bf620eb6da66a

ubergesundheit replied:
It looks like `m2r2` is not working in the new image

```
$ make -C Documentation update-helm-values
make: Entering directory '/home/gerald/code/work/cilium-upstream/Documentation'
docker container run --rm --workdir /src/Documentation --volume /home/gerald/code/work/cilium-upstream/Documentation/..:/src --env READTHEDOCS_VERSION= --env SKIP_LINT= --env INCREMENTAL= --user "1000:1000" "quay.io/cilium/helm-toolbox:"v1.1.1"@sha256:"722901c26a9edb75a514927f9f759302a0dc03edacadc835a76bf620eb6da66a"" helm-docs -d -c /src/install/kubernetes -t /src/Documentation/helm-values.tmp.tmpl > helm-values.tmp
time="2024-03-07T06:49:24Z" level=info msg="Found Chart directories [cilium]"
time="2024-03-07T06:49:24Z" level=info msg="Generating README Documentation for chart /src/install/kubernetes/cilium"
awk -F'|' '{print "|"$2"|"$5"|"$3"|"$4"|"}' helm-values.tmp > helm-values.awk
docker container run --rm --workdir /src/Documentation --volume /home/gerald/code/work/cilium-upstream/Documentation/..:/src --env READTHEDOCS_VERSION= --env SKIP_LINT= --env INCREMENTAL= --user "1000:1000" "quay.io/cilium/helm-toolbox:"v1.1.1"@sha256:"722901c26a9edb75a514927f9f759302a0dc03edacadc835a76bf620eb6da66a"" python3 /usr/bin/m2r2 --overwrite helm-values.awk
Traceback (most recent call last):
  File "/usr/bin/m2r2", line 5, in <module>
    from m2r2 import main
ModuleNotFoundError: No module named 'm2r2'
make: *** [Makefile:106: helm-values.rst] Error 1
make: Leaving directory '/home/gerald/code/work/cilium-upstream/Documentation'
```

joestringer replied:
Alright, we can stick to 1.1.0 for now and follow up separately on this.

ubergesundheit commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1746,6 +1921,9 @@ nodePort:
 # default, always and never.
 # ref: https://docs.cilium.io/en/stable/security/policy/intro/#policy-enforcement-modes
 policyEnforcementMode: "default"
+# @schema
+# type: [null, string, array]
+# @schema
 # -- policyCIDRMatchMode is a list of entities that may be selected by CIDR selector.
 # The possible value is "nodes".
 policyCIDRMatchMode:
```
This option seems to be defined as [`StringSlice`](https://github.com/cilium/cilium/blob/fc1f87bd2285778c7fb582fa23376ee8d52d61cd/daemon/cmd/daemon_main.go#L1069) in the daemon Go code. The helm template also uses `join` which usually operates on StringSlices. Sprig (templating functions library used by helm) [casts the given string to a string slice with length 1](https://github.com/Masterminds/sprig/blob/581758eb7d96ae4d113649668fa96acc74d46e7f/strings.go#L159). I guess thats the reason why both arrays and single strings seem to work in this case.

Validation wise this seems okay but might turn heads of curious readers.

Same goes for `ipam.operator.clusterPoolIPv4PodCIDRList` and possibly more

gandro replied:
It is a bit unfortunate, but I'm pretty sure that some users will have `clusterPoolIPv4PodCIDRList` set to a string, because it's predecessor value used to be a string too.

Ideally we'd want to deprecate the uses of strings here - but on the other hand, we want to be able to do it incrementally. It is a bit annoying that the Helm schema is "all or nothing" and that there is not a "emit a warning if a unsupported type or additional property is used" mode that would allow us to deprecate and fix those inconsistencies.

In any case, this seems fine as is for now.

gandro reviewed (approved):
Re-reviewed, this looks good to me now, assuming it will pass CI! Thanks a lot

chancez reviewed (approved):
Overall, I think the change looks good. I tested with a bunch of the `values.yaml` files I have locally and didn't hit any issues. I don't think there's anything that should necessarily prevent it from merging, especially since I know there's a high likelihood of merge conflicts the longer it waits, but there's a few things I think we should probably consider, either now, or in a follow up PR:

- We should probably use `$ref` for well-known types. Eg: `affinity`, `antiAffinity`, `tolerations`, `securityContext`, `topologySpreadConstraints`, etc. For example: https://github.com/apache/airflow/blob/2cc2f5e7d1307a02c5ed9cd96e7c5688dae622b1/chart/values.schema.json#L174-L209
- Eventually we'll want to expand the validation introduced here, taking advantage of the ability to validate the values themselves, in addition to their types. Should we try any of the other features out (eg: validating enum values) to verify `helm-schema` will work for those use cases? I'm not suggest we do this for a lot of fields, just a couple to test each feature we would expect to use.
- What's the path forward if we wanted to enable stricter validation, ie: disallowing unknown fields? IIRC we would need to use `additionalProperties` somehow. Is that what the `--skip-auto-generation title,description,required,default,additionalProperties` is doing? 

Again, none of these are blockers, but I do want to make sure we're looking slightly ahead, as it's likely we will want to take advantage of this schema validation more and more in the future, once this is merged, so it should be robust enough to be able to do those things.

sayboras reviewed (approved):
Thanks and LGTM :heavy_check_mark:
